### PR TITLE
fix: improve typing of RuleResult

### LINKED ISSUE #352: RuleResult class missing from type declarations
`RuleResult` has a `toJSON()` method ([source](https://github.com/CacheControl/json-rules-engine/blob/master/src/rule-result.js#L36))

The `RuleResult` interface declaration does not include this ([source](https://github.com/CacheControl/json-rules-engine/blob/master/types/index.d.ts#L127)). Also, there is no class declaration for RuleResult.

The impact of this issue is felt when handling rule results, in my case decomposing them from an EngineResult. I ran into the same problem outlined in https://github.com/CacheControl/json-rules-engine/issues/172 which was "resolved" with the introduction of the `toJSON` method to `RuleResult`s. This resolution doesn't help if you're using typescript, as the compiler has no knowledge of this method.

Screenshot 2023-09-14 at 09 54 49

To make my code work around this, I've had to create a new `RuleResult` type in my code and extend it to include the `toJSON` method, which is a bit yuck.

A good solution would include:

1. add and export a `RuleResult` class declaration which includes the `toJSON()` method
2. add and export a `RuleResultSerializable` type declaration (taking inspiration from the type declaration for [`Rule` class declaration](https://github.com/CacheControl/json-rules-engine/blob/master/types/index.d.ts#L144-L148))

Thanks

### PR DESCRIPTION
**Pull request by emilefokkemanavara**
This PR closes #352 by improving the typing of [RuleResult](https://github.com/CacheControl/json-rules-engine/blob/ae8fa0553968ce3ac2dd5dfa9508ce208a852047/types/index.d.ts#L156).

### CODE REVIEW

**Comment by chris-pardy** [comment]
@emilefokkemanavara thanks for putting this up. I'll take a look later today or over the weekend.
One thing to note right away is that the condition reference can end up in the result if the conditions it's referencing isn't defined and if the "allowUndefinedConditions" option is specified.

**Comment by emilefokkemanavara** [comment]
@chris-pardy Thanks for the heads up!

File: types/index.d.ts [hunk 1]
```diff
@@ -184,6 +184,14 @@ export class Rule implements RuleProperties {
   ): T extends true ? string : RuleSerializable;
 }
 
+interface BooleanConditionResultProperties {
+  result: boolean
+}
+
+interface ConditionResultProperties extends BooleanConditionResultProperties {
+  factResult: any
```
Comment (chris-pardy) [inline types/index.d.ts:194]: generally we should start using `unknown` for values like this rather than `any`

Comment (emilefokkemanavara) [reply types/index.d.ts]: done!

**Comment by emilefokkemanavara** [comment]
> @emilefokkemanavara This looks great, expect for one issue.
> 
> The a evaluation of JSON Rules Engine follows these steps:
> 
> 1. Clone all the conditions
> 2. Attach the cloned conditions to the rule result
> 3. Evaluate each cloned condition
> 4. Modify the cloned condition to have new properties such as `result`
> 
> This means that if a condition doesn't get evaluated it will still be in the rule result but it won't have properties like `result` or `factResult`. The main reason conditions can be skipped is in `Any`/`All` conditions. These group by priority and short-circuit between priorities so for instance:
> 
> This set of conditions
> 
> ```json
> {
>   "all": [
>      { "fact": "age", "operator": "greaterThanInclusive", "value": 21, "priority": 1 },
>      { "fact": "likesOranges", "operator": "equal", "value": true, "priority": 2 }
>   ]
> }
> ```
> 
> The result can look like this.
> 
> ```json
> {
>   "all": [
>      {"fact": "age", "operator": "greaterThanInclusive", "value": 21, "priority": 1, "factResult": 20 },
>      { "fact": "likesOranges", "operator": "equal", "value": true, "priority": 2 }
>   ]
> }
> ```

@chris-pardy So this means `result` and `factResult` are possibly undefined? (see latest commit)

**Review by chris-pardy — COMMENTED** [review]
Looks good.
Before running the tests could you rebase this and squash the commits to make it a bit easier to read down the line.

**Comment by chris-pardy** [comment]
> > @emilefokkemanavara This looks great, expect for one issue.
> > 
> > The a evaluation of JSON Rules Engine follows these steps:
> > 
> > 1. Clone all the conditions
> > 2. Attach the cloned conditions to the rule result
> > 3. Evaluate each cloned condition
> > 4. Modify the cloned condition to have new properties such as `result`
> > 
> > This means that if a condition doesn't get evaluated it will still be in the rule result but it won't have properties like `result` or `factResult`. The main reason conditions can be skipped is in `Any`/`All` conditions. These group by priority and short-circuit between priorities so for instance:
> > 
> > This set of conditions
> > 
> > ```json
> > {
> >   "all": [
> >      { "fact": "age", "operator": "greaterThanInclusive", "value": 21, "priority": 1 },
> >      { "fact": "likesOranges", "operator": "equal", "value": true, "priority": 2 }
> >   ]
> > }
> > ```
> > 
> > The result can look like this.
> > 
> > ```json
> > {
> >   "all": [
> >      {"fact": "age", "operator": "greaterThanInclusive", "value": 21, "priority": 1, "factResult": 20 },
> >      { "fact": "likesOranges", "operator": "equal", "value": true, "priority": 2 }
> >   ]
> > }
> > ```
> 
> @chris-pardy So this means `result` and `factResult` are possibly undefined? (see latest commit)

Yes, that's correct.

File: types/index.d.ts [hunk 2]
```diff
@@ -186,25 +204,46 @@ interface ConditionProperties {
...
+  any: NestedConditionResult[]
+} & BooleanConditionResultProperties
 type NotConditions = { not: NestedCondition; name?: string; priority?: number };
+type NotConditionsResult = NotConditions & {not: NestedConditionResult} & BooleanConditionResultProperties;
 type ConditionReference = {
   condition: string;
   name?: string;
   priority?: number;
 };
+type ConditionReferenceResult = ConditionReference & BooleanConditionResultProperties
```
[the remark this answers is not in the document]
Comment (emilefokkemanavara) [reply types/index.d.ts:234]: [There's the false](https://github.com/CacheControl/json-rules-engine/blob/ae8fa0553968ce3ac2dd5dfa9508ce208a852047/src/rule.js#L329) if the reference is undefined.

**Review by chris-pardy — APPROVED** [review]

**Pull request merged**